src/include/bufferpool.h

```diff
@@ -0,0 +1,162 @@
+
+            /** Return true if the specified transaction has a lock on the specified page */
+            bool HoldsLock(TransactionId* tid, Page);
+
+            /**
+             * Commit or abort a given transaction; release all locks associated to
+             * the transaction.
+             *
+             * @param tid the ID of the transaction requesting the unlock
+             * @param commit a flag indicating whether we should commit or abort
+             */
+            void TransactionComplete(TransactionId* tid, bool commit); //IO exception
+
+            /**
+             * Add a tuple to the specified table behalf of transaction tid.  Will
+             * acquire a write lock on the page the tuple is added to(Lock 
+             * acquisition is not needed for lab2). May block if the lock cannot 
+             * be acquired.
+             * 
+             * Marks any pages that were dirtied by the operation as dirty by calling
+             * their markDirty bit, and updates cached versions of any pages that have 
+             * been dirtied so that future requests see up-to-date pages. 
+             *
+             * @param tid the transaction adding the tuple
+             * @param tableId the table to add the tuple to
+             * @param t the tuple to add
+             */
+            void InsertTuple(TransactionId* tid, int tableId, Tuple* t); // throws DbException, IOException, TransactionAbortedException
+
+            /**
+             * Remove the specified tuple from the buffer pool.
+             * Will acquire a write lock on the page the tuple is removed from. May block if
+             * the lock cannot be acquired.
+             *
+             * Marks any pages that were dirtied by the operation as dirty by calling
+             * their markDirty bit.  Does not need to update cached versions of any pages that have 
+             * been dirtied, as it is not possible that a new page was created during the deletion
+             * (note difference from addTuple).
+             *
+             * @param tid the transaction deleting the tuple.
+             * @param t the tuple to delete
+             */
+            void DeleteTuple(TransactionId* tid, Tuple* t);
+
+            /**
+             * Flush all dirty pages to disk.
+             * NB: Be careful using this routine -- it writes dirty data to disk so will
+             *     break simpledb if running in NO STEAL mode.
+             */
+            void FlushAllPages();
+
+            /** Remove the specific page id from the buffer pool.
+                Needed by the recovery manager to ensure that the
+                buffer pool doesn't keep a rolled back page in its
+                cache.
+            */
+            void DiscardPage(PageId* pid);
+
+            /**
+             * Flushes a certain page to disk
+             * @param pid an ID indicating the page to flush
+             */
+            void FlushPage(PageId* pid);
```

Comment:
You already correctly declared this function in the private section above, so just move the function specification comments up and remove this declaration!